### PR TITLE
1902: upgrade `CLI11` to `v2.2.0`

### LINKED ISSUE #1902: Upgrade CLI11 library
**What Needs to be Done?**
Upgrade CLI11 library to the latest version and make sure that it works with C++17.

### PR DESCRIPTION
**Pull request by cz4rs**
Upgrade `CLI11` to the latest release.
Note: this keeps the added `namespace vt` in the header.

fixes #1902

### CODE REVIEW

File: src/vt/configs/arguments/args.cc [hunk 1]
```diff
@@ -484,32 +484,24 @@ void addLbArgs(CLI::App& app, AppConfig& appConfig) {
...
-  auto lbqq = "";
   auto s  = app.add_flag("--vt_lb", appConfig.vt_lb, lb);
   auto t1 = app.add_flag("--vt_lb_quiet", appConfig.vt_lb_quiet, lb_quiet);
-  auto u  = app.add_option("--vt_lb_file_name", appConfig.vt_lb_file_name, lb_file_name, lbf)->check(CLI::ExistingFile);
+  auto u  = app.add_option("--vt_lb_file_name", appConfig.vt_lb_file_name, lb_file_name)->capture_default_str()->check(CLI::ExistingFile);
   auto u1 = app.add_flag("--vt_lb_show_spec", appConfig.vt_lb_show_spec, lb_show_spec);
-  auto v  = app.add_option("--vt_lb_name", appConfig.vt_lb_name, lb_name, lbn);
-  auto v1 = app.add_option("--vt_lb_args", appConfig.vt_lb_args, lb_args, lba);
-  auto w  = app.add_option("--vt_lb_interval", appConfig.vt_lb_interval, lb_interval, lbi);
+  auto v  = app.add_option("--vt_lb_name", appConfig.vt_lb_name, lb_name)->capture_default_str();
```
Comment (cz4rs) [inline src/vt/configs/arguments/args.cc:491]: New feature reduces the code duplication: default value will be read from `AppConfig::vt_lb_name`.

**Review by PhilMiller — COMMENTED** [review]
This all looks very nice. 

Could you confirm that all of the defaults actually matched?

**Review by nmm0 — APPROVED** [review]
Looks good to me, also fixes C++17 compile issue

**Review by PhilMiller — CHANGES_REQUESTED** [review]
Please work around the Intel CI failure before merging, so that we don't just see that failure on every PR going forward

**Comment by PhilMiller** [comment]
```
FAILED: src/CMakeFiles/vt.dir/Unity/unity_0_cxx.cxx.o 
/usr/bin/ccache /usr/local/bin/mpicxx -DJSON_USE_IMPLICIT_CONVERSIONS=1 -DVT_NO_COLOR_ENABLED -I/vt/lib/CLI -I/build/vt/release -I/vt/src -I/vt/lib/json/include -I/vt/lib/brotli/c/include -I/vt/lib/libfort/lib -isystem /vt/lib/fmt/include -isystem /vt/lib/EngFormat-Cpp/include -isystem /build/checkpoint/install/include -isystem /build/detector/install/include -O3 -DNDEBUG -Wall -pedantic -Wshadow -Wno-unknown-pragmas -Wsign-compare -Werror -fPIC -std=c++14 -MD -MT src/CMakeFiles/vt.dir/Unity/unity_0_cxx.cxx.o -MF src/CMakeFiles/vt.dir/Unity/unity_0_cxx.cxx.o.d -o src/CMakeFiles/vt.dir/Unity/unity_0_cxx.cxx.o -c /build/vt/src/CMakeFiles/vt.dir/Unity/unity_0_cxx.cxx
In file included from /vt/src/vt/configs/arguments/args.cc(54),
                 from /build/vt/src/CMakeFiles/vt.dir/Unity/unity_0_cxx.cxx(9):
/vt/lib/CLI/CLI/CLI11.hpp(8143): error #3280: declaration hides variable "optarg" (declared at line 36 of "/usr/include/x86_64-linux-gnu/bits/getopt_core.h")
                      std::string optarg = args.back();
                                  ^
```

File: lib/CLI/CLI/CLI11.hpp [hunk 2]
```diff
@@ -8140,9 +8140,9 @@ class App {
                     break;
                 }
                 if(validate_optional_arguments_) {
-                    std::string optarg = args.back();
-                    optarg = op->_validate(optarg, 0);
-                    if(!optarg.empty()) {
+                    std::string cli_optarg = args.back();
```
Comment (cz4rs) [inline lib/CLI/CLI/CLI11.hpp:8143]: I think that's better than adding `Wno-shadow` for icpc.
Btw this is already fixed at `main` branch of CLI11.

File: src/vt/configs/arguments/args.cc [hunk 3]
```diff
@@ -230,13 +230,13 @@ void addMemUsageArgs(CLI::App& app, AppConfig& appConfig) {
   auto mem_thresh    = "The threshold increments to print memory usage: \"<value> {GiB,MiB,KiB,B}\"";
   auto mem_sched     = "The frequency to query the memory threshold check (some memory reporters might be expensive)";
   auto mem_footprint = "Print live components' memory footprint after initialization and before shutdown";
-  auto mm = app.add_option("--vt_memory_reporters",          appConfig.vt_memory_reporters, mem_desc, true);
+  auto mm = app.add_option("--vt_memory_reporters",          appConfig.vt_memory_reporters, mem_desc)->capture_default_str();
```
Comment (cz4rs) [inline src/vt/configs/arguments/args.cc:233]: `true` here meant that the default value was already captured.

File: src/vt/configs/arguments/args.cc [hunk 4]
```diff
@@ -264,9 +264,9 @@ void addStackDumpArgs(CLI::App& app, AppConfig& appConfig) {
   auto h = app.add_flag("--vt_no_assert_stack", appConfig.vt_no_assert_stack, assert);
   auto i = app.add_flag("--vt_no_abort_stack",  appConfig.vt_no_abort_stack,  abort);
   auto j = app.add_flag("--vt_no_stack",        appConfig.vt_no_stack,        stack);
-  auto k = app.add_option("--vt_stack_file",    appConfig.vt_stack_file,      file, "");
-  auto l = app.add_option("--vt_stack_dir",     appConfig.vt_stack_dir,       dir,  "");
-  auto m = app.add_option("--vt_stack_mod",     appConfig.vt_stack_mod,       mod,  1);
+  auto k = app.add_option("--vt_stack_file",    appConfig.vt_stack_file,      file)->capture_default_str();
+  auto l = app.add_option("--vt_stack_dir",     appConfig.vt_stack_dir,       dir)->capture_default_str();
+  auto m = app.add_option("--vt_stack_mod",     appConfig.vt_stack_mod,       mod)->capture_default_str();
```
Comment (cz4rs) [inline src/vt/configs/arguments/args.cc:269]: Difference:
https://github.com/DARMA-tasking/vt/blob/e8ca3b556b5cb98d82b92c31b75cd084ddc799b4/src/vt/configs/arguments/app_config.h#L117

File: src/vt/configs/arguments/args.cc [hunk 5]
```diff
@@ -302,18 +302,18 @@ void addTraceArgs(CLI::App& app, AppConfig& appConfig) {
   auto n  = app.add_flag("--vt_trace",                   appConfig.vt_trace,                   trace);
   auto nm = app.add_option("--vt_trace_mpi",             arg_trace_mpi,                      trace_mpi)
     ->check(CLI::IsMember({"internal", "external"}));
-  auto o  = app.add_option("--vt_trace_file",            appConfig.vt_trace_file,              tfile,     "");
-  auto p  = app.add_option("--vt_trace_dir",             appConfig.vt_trace_dir,               tdir,      "");
-  auto q  = app.add_option("--vt_trace_mod",             appConfig.vt_trace_mod,               tmod,      1);
-  auto qf = app.add_option("--vt_trace_flush_size",      appConfig.vt_trace_flush_size,        tflushmod, 0);
+  auto o  = app.add_option("--vt_trace_file",            appConfig.vt_trace_file,              tfile)->capture_default_str();
+  auto p  = app.add_option("--vt_trace_dir",             appConfig.vt_trace_dir,               tdir)->capture_default_str();
+  auto q  = app.add_option("--vt_trace_mod",             appConfig.vt_trace_mod,               tmod)->capture_default_str();
```
Comment (cz4rs) [inline src/vt/configs/arguments/args.cc:307]: Difference:
https://github.com/DARMA-tasking/vt/blob/e8ca3b556b5cb98d82b92c31b75cd084ddc799b4/src/vt/configs/arguments/app_config.h#L129

File: src/vt/configs/arguments/args.cc [hunk 6]
```diff
@@ -484,32 +484,24 @@ void addLbArgs(CLI::App& app, AppConfig& appConfig) {
...
   auto wl = app.add_flag("--vt_lb_keep_last_elm", appConfig.vt_lb_keep_last_elm, lb_keep_last_elm);
   auto ww = app.add_flag("--vt_lb_data", appConfig.vt_lb_data, lb_data);
   auto xz = app.add_flag("--vt_lb_data_compress", appConfig.vt_lb_data_compress, lb_data_comp);
-  auto wx = app.add_option("--vt_lb_data_dir", appConfig.vt_lb_data_dir, lb_data_dir, lbd);
-  auto wy = app.add_option("--vt_lb_data_file", appConfig.vt_lb_data_file, lb_data_file,lbs);
-  auto xx = app.add_option("--vt_lb_data_dir_in", appConfig.vt_lb_data_dir_in, lb_data_dir_in, lbd);
-  auto xy = app.add_option("--vt_lb_data_file_in", appConfig.vt_lb_data_file_in, lb_data_file_in, lbs);
+  auto wx = app.add_option("--vt_lb_data_dir", appConfig.vt_lb_data_dir, lb_data_dir)->capture_default_str();
+  auto wy = app.add_option("--vt_lb_data_file", appConfig.vt_lb_data_file, lb_data_file)->capture_default_str();
+  auto xx = app.add_option("--vt_lb_data_dir_in", appConfig.vt_lb_data_dir_in, lb_data_dir_in)->capture_default_str();
```
Comment (cz4rs) [inline src/vt/configs/arguments/args.cc:499]: https://github.com/DARMA-tasking/vt/blob/e8ca3b556b5cb98d82b92c31b75cd084ddc799b4/src/vt/configs/arguments/app_config.h#L149

File: src/vt/configs/arguments/args.cc [hunk 7]
```diff
@@ -484,32 +484,24 @@ void addLbArgs(CLI::App& app, AppConfig& appConfig) {
...
+  auto xy = app.add_option("--vt_lb_data_file_in", appConfig.vt_lb_data_file_in, lb_data_file_in)->capture_default_str();
```
Comment (cz4rs) [inline src/vt/configs/arguments/args.cc:500]: https://github.com/DARMA-tasking/vt/blob/e8ca3b556b5cb98d82b92c31b75cd084ddc799b4/src/vt/configs/arguments/app_config.h#L151

**Comment by cz4rs** [comment]
> This all looks very nice.
> 
> Could you confirm that all of the defaults actually matched?

I have double-checked and noted the differences in the comments.

**Comment by cz4rs** [comment]
Rebased on top of `develop`.
All the differences regarding default values should only make the `--vt_help` output more accurate (the ones in `args.cc` were out of sync).

**Review by PhilMiller — APPROVED** [review]

**Pull request merged**